### PR TITLE
core: (irdl) Add regions to declarative assembly format

### PR DESCRIPTION
**Pull request by kimxworrall**
Added support for regions to the declarative assembly format.

The declarative assembly format now supports:
- Referring to region variables
- Referring to optional or variadic region variables
- Printing custom assembly for region variables

The format requires that:
- The attr-dict-with-keyword directive is present and not attr-dict. This is to maintain consistency with mlir.
- The regions appear after the attributes.

Also added tests.

### CODE REVIEW

File: xdsl/irdl/declarative_assembly_format.py [hunk 1]
```diff
@@ -18,7 +20,9 @@
     ParametrizedAttribute,
     SSAValue,
     TypedAttribute,
+    Region,
 )
+from xdsl.ir.core import Block, Operation
```
Comment (alexarice) [inline xdsl/irdl/declarative_assembly_format.py:25]: Please install the precommit, this shouldn't pass the linter

**Review by alexarice — COMMENTED** [review]
Could you comment on why regions must come after attributes? I can't see anything about this in the mlir documentation (which doesn't mean it's not true)

File: xdsl/irdl/declarative_assembly_format.py (hunk 1 above)
Comment (kimxworrall) [reply xdsl/irdl/declarative_assembly_format.py:25]: That's the part I'm not completely sure about from the documentation, I know that the generic format does this, and that the attribute dict has to be attr-dict-with-keyword, but am unsure if its allowed to have the attributes in the middle of regions (but labelled).

Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py:25]: I tried some stuff locally and the format "$body attr-dict-with-keyword" seems to work absolutely fine

File: .github/workflows/update_xdsl_pyodide_build.py
Comment (superlopuh) [inline .github/workflows/update_xdsl_pyodide_build.py:1]: why this change?

File: tests/irdl/test_declarative_assembly_format.py [hunk 2]
```diff
@@ -1087,6 +1090,194 @@ class OptionalResultOp(IRDLOperation):
...
+    """Test the parsing of multiple optional regions"""
+
+    """Test that multiple optional regions requires the ABCMeta PyRDL option."""
+    with pytest.raises(
+        PyRDLOpDefinitionError,
+        match="Operation test.optional_regions defines more than two variadic regions",
+    ):
+
+        @irdl_op_definition
+        class OptionalOperandsOp(IRDLOperation):
```
Comment (superlopuh) [inline tests/irdl/test_declarative_assembly_format.py:1273]:
```suggestion
        class OptionalOperandsOp(IRDLOperation):  # pyright: ignore[reportUnusedClass]
```

Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:1273]: Surely the warning shouldn't be ignored and is an indication that something is wrong?

File: xdsl/irdl/declarative_assembly_format.py [hunk 3]
```diff
@@ -159,11 +163,25 @@ def parse(
...
+        regions = cast(
+            Sequence[
+                Region
+                | Sequence[Operation]
+                | Sequence[Block]
+                | Sequence[Region | Sequence[Operation] | Sequence[Block]]
+                | None
+            ],
+            state.regions,
+        )
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format.py:177]: I'm surprised this is needed, I'll clone and play with this locally quickly

File: xdsl/irdl/operations.py [hunk 4]
```diff
@@ -1735,7 +1735,7 @@ def fun(self: Any, idx: int = arg_idx, previous_vars: int = previous_variadics):
         not any(isinstance(o, arg_size_option) for o in op_def.options)
     ):
         arg_size_option_name = type(arg_size_option).__name__
-        raise Exception(
+        raise PyRDLOpDefinitionError(
```
Comment (superlopuh) [inline xdsl/irdl/operations.py:1738]: 👍

File: xdsl/tools/xdsl-opt
Comment (superlopuh) [inline xdsl/tools/xdsl-opt:1]: why this change?

File: tests/irdl/test_declarative_assembly_format.py (hunk 2 above)
Comment (alexarice) [reply tests/irdl/test_declarative_assembly_format.py:1273]: Ignore me, didn't see this was just a test

File: tests/irdl/test_declarative_assembly_format.py [hunk 5]
```diff
@@ -1087,6 +1090,194 @@ class OptionalResultOp(IRDLOperation):
...
+        @irdl_op_definition
+        class NoRegionTypeOp(IRDLOperation):  # pyright: ignore[reportUnusedClass]
+            name = "test.region_op_missing_keyword"
+            region = region_def()
+
+            assembly_format = "attr-dict $region"
+
+
+def test_region_before_attr():
+    """Test that regions should appear after the 'attr-dict-with-keyword' directive."""
```
Comment (alexarice) [inline tests/irdl/test_declarative_assembly_format.py:1126]: My mlir testing suggests that this should be valid assembly format

File: xdsl/irdl/declarative_assembly_format.py (hunk 1 above)
Comment (kimxworrall) [reply xdsl/irdl/declarative_assembly_format.py]: Hmm okay, maybe its because I'm running an older version, will fix

File: xdsl/irdl/declarative_assembly_format.py [hunk 6]
```diff
@@ -723,6 +741,80 @@ def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> No
...
+        state.should_emit_space = True
+
+
+@dataclass(frozen=True)
+class VariadicRegionVariable(
+    VariadicVariable, VariableDirective, OptionallyParsableDirective
+):
+    """
+    A variadic region variable, with the following format:
+      region-directive ::= ( percent-ident ( `,` percent-id )* )?
```
Comment (alexarice) [inline xdsl/irdl/declarative_assembly_format.py:770]: Is this the correct format? SHould it not still be `region-directive ::= dollar-ident`?

File: xdsl/irdl/declarative_assembly_format.py [hunk 7]
```diff
@@ -723,6 +741,80 @@ def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> No
             state.should_emit_space = True
 
 
+@dataclass(frozen=True)
+class RegionVariable(VariableDirective):
```
Comment (alexarice) [inline xdsl/irdl/declarative_assembly_format.py:745]: Surely this can be `OptionallyParsable` too

File: xdsl/irdl/declarative_assembly_format.py [hunk 8]
```diff
@@ -723,6 +741,80 @@ def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> No
...
+        if region:
+            printer.print_list(region, printer.print_region, delimiter=" ")
+            state.last_was_punctuation = False
+            state.should_emit_space = True
+
+
+class OptionalRegionVariable(OptionalVariable, OptionallyParsableDirective):
+    """
+    An optional region variable, with the following format:
+      region-directive ::= ( percent-ident )?
```
Comment (alexarice) [inline xdsl/irdl/declarative_assembly_format.py:797]: This format seems incorrect again (see above)?

**Review by alexarice — COMMENTED** [review]
Haven't read through everything but here's some comments

File: xdsl/irdl/declarative_assembly_format.py (hunk 1 above)
Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py]: From my testing it really only seems to complain if you have an attr-dict immediately followed by a region

File: xdsl/tools/xdsl-opt
Comment (kimxworrall) [reply xdsl/tools/xdsl-opt:1]: Ah I think this is just the cache auto-updated and it overwrote the file without changing anything.

File: xdsl/irdl/declarative_assembly_format.py (hunk 1 above)
Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py]: https://github.com/llvm/llvm-project/blob/ea28668f8c980c9987b0e2f305b0a8eff9797401/mlir/tools/mlir-tblgen/OpFormatGen.cpp#L3010 Here is the check

Comment (kimxworrall) [reply xdsl/irdl/declarative_assembly_format.py]: As in does not accept "attr-dict $body" but does accept "attr-dict-with-keyword $body"?

Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py]: Yes, it looks like the only thing that is denied is attr-dict followed directly by a region variable

Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py]: I also tested ```attr-dict `,` $body``` and that worked as expected too.

File: xdsl/irdl/declarative_assembly_format.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py:177]: https://github.com/xdslproject/xdsl/commit/a7ba1d417b16f0386e4a019704ea750dcdb43715

Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py:177]: I don't think the `None` in the inner `list[Region | None]` is needed, might as well remove?

Comment (kimxworrall) [reply xdsl/irdl/declarative_assembly_format.py:177]: Thank you, that seems sensible

Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py:177]: Feel free to merge those changes into this PR, then I think this is GTG

File: xdsl/tools/xdsl-opt
Comment (superlopuh) [inline xdsl/tools/xdsl-opt:1]: can revert?

File: xdsl/irdl/declarative_assembly_format.py (hunk 3 above)
Comment (kimxworrall) [reply xdsl/irdl/declarative_assembly_format.py]: Thanks, have done so now

File: xdsl/irdl/declarative_assembly_format.py (hunk 1 above)
Comment (kimxworrall) [reply xdsl/irdl/declarative_assembly_format.py]: Have made these updates and added tests, should work now

**Review by alexarice — COMMENTED** [review]
There's still two files that you've not touched being pulled in somehow. I've left a few minor comments but looking good, will give it a test tomorrow

**Comment by alexarice** [comment]
I also have no idea why the mlir integration tests are failing

File: xdsl/irdl/declarative_assembly_format.py [hunk 9]
```diff
@@ -48,19 +49,20 @@ class ParsingState:
...
     properties: dict[str, Attribute]
     constraint_context: ConstraintContext
 
     def __init__(self, op_def: OpDef):
-        if op_def.regions or op_def.successors:
+        if op_def.successors:
             raise NotImplementedError(
-                "Operation definitions with regions "
-                "or successors are not yet supported"
+                "Operation definitions with " "successors are not yet supported"
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py:60]:
```suggestion
                "Operation definitions with successors are not yet supported"
```

File: xdsl/irdl/declarative_assembly_format.py [hunk 10]
```diff
@@ -159,11 +161,13 @@ def parse(
             properties = state.properties
         else:
             properties = op_def.split_properties(state.attributes)
+
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format.py:164]:
```suggestion
```

Comment (superlopuh) [reply xdsl/irdl/declarative_assembly_format.py:164]: I feel like it's easier just to post suggestions then it's two clicks in the GH UI to apply them

Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py:164]: I just forgot this was a feature

File: xdsl/irdl/declarative_assembly_format.py [hunk 11]
```diff
@@ -723,6 +727,85 @@ def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> No
...
 
+@dataclass(frozen=True)
+class RegionVariable(VariableDirective, OptionallyParsableDirective):
+    """
+    A region variable, with the following format:
+      region-directive ::= dollar-ident
+    The directive will request a space to be printed after.
+    """
+
+    def parse(self, parser: Parser, state: ParsingState) -> None:
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py:737]: Oh wait: Only the OptionalRegionVariabl needs to implement it, not this one, indeed :+1: Good catch!

File: xdsl/irdl/declarative_assembly_format.py [hunk 12]
```diff
@@ -723,6 +727,85 @@ def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> No
...
+        state.should_emit_space = True
+
+
+@dataclass(frozen=True)
+class VariadicRegionVariable(
+    VariadicVariable, VariableDirective, OptionallyParsableDirective
+):
+    """
+    A variadic region variable, with the following format:
+      region-directive ::= ( dollar-ident ( `,` dollar-id )* )?
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py:760]: Agreed, it should just be dollar-ident.

File: xdsl/irdl/declarative_assembly_format.py [hunk 13]
```diff
@@ -723,6 +726,85 @@ def print(self, printer: Printer, state: PrintingState, op: IRDLOperation) -> No
...
+        if region:
+            printer.print_list(region, printer.print_region, delimiter=" ")
+            state.last_was_punctuation = False
+            state.should_emit_space = True
+
+
+class OptionalRegionVariable(OptionalVariable, OptionallyParsableDirective):
+    """
+    An optional region variable, with the following format:
+      region-directive ::= ( dollar-ident )?
```
Comment (PapyChacal) [inline xdsl/irdl/declarative_assembly_format.py:787]: Same here, should just be dollar-ident

File: tests/irdl/test_declarative_assembly_format.py [hunk 14]
```diff
@@ -1087,6 +1090,213 @@ class OptionalResultOp(IRDLOperation):
...
+    check_roundtrip(program, ctx)
+    check_equivalence(program, generic_program, ctx)
+
+
+@irdl_op_definition
+class MiscOp(IRDLOperation):
+    name = "test.typed_attr"
+    attr = attr_def(IntegerAttr[I32])
+
+    assembly_format = "$attr attr-dict"
```
Comment (PapyChacal) [inline tests/irdl/test_declarative_assembly_format.py:1163]: Minor: could you maybe populate the below tests with test.op (TestOp) instead of creating a new one

File: tests/irdl/test_declarative_assembly_format.py [hunk 15]
```diff
@@ -1087,6 +1090,213 @@ class OptionalResultOp(IRDLOperation):
...
+        (
+            "attr-dict-with-keyword $region",
+            'test.variadic_region {\n  "test.op"() : () -> ()\n} {\n  "test.op"() : () -> ()\n}',
+            '"test.variadic_region"() ({ "test.op"() : () -> ()}, { "test.op"() : () -> ()}) : () -> ()',
+        ),
+        (
+            "attr-dict-with-keyword $region",
+            'test.variadic_region {\n  "test.op"() : () -> ()\n} {\n  "test.op"() : () -> ()\n}',
+            '"test.variadic_region"() ({ "test.op"() : () -> ()}, {"test.op"() : () -> ()}) : () -> ()',
+        ),
```
Comment (PapyChacal) [inline tests/irdl/test_declarative_assembly_format.py:1228]: Those two tests are the same!
I'm fine just keeping one, or making the last one a 3-regions example for the show

File: tests/irdl/test_declarative_assembly_format.py [hunk 16]
```diff
@@ -1087,6 +1090,213 @@ class OptionalResultOp(IRDLOperation):
...
+    ctx.load_dialect(Test)
+
+    check_roundtrip(program, ctx)
+    check_equivalence(program, generic_program, ctx)
+
+
+def test_multiple_optional_regions():
+    """Test the parsing of multiple optional regions"""
+
+    """Test that multiple optional regions requires the ABCMeta PyRDL option."""
```
Comment (PapyChacal) [inline tests/irdl/test_declarative_assembly_format.py:1285]: This just requires the right IRDLOption, and I believe this is what this test tests, and it shouldn't be here.
The declarative format framework should check that two variadic regions do not directly follow each other (ambiguous). It might already through baseclasses, but here it catches the irdl_op_definition even before that.

File: tests/irdl/test_declarative_assembly_format.py [hunk 17]
```diff
@@ -1087,6 +1090,213 @@ class OptionalResultOp(IRDLOperation):
...
+
+        @irdl_op_definition
+        class OptionalOperandsOp(IRDLOperation):  # pyright: ignore[reportUnusedClass]
+            name = "test.optional_regions"
+            region1 = opt_region_def()
+            region2 = opt_region_def()
+
+            assembly_format = "attr-dict-with-keyword $region1 $region2"
+
+
```
Comment (PapyChacal) [inline tests/irdl/test_declarative_assembly_format.py:1299]: You implemented it in such a nice way that it should work with optional groups, maybe we wont to test those?

Things like `` ($opt_region^ `keyword`)? `` and `` (`keyword` $opt_region^)? ``

Don't take this as a requirement, would only be even nicer if you have time for it too, but I'd be glad to merge it without that too!

**Review by PapyChacal — APPROVED** [review]
Amazing to see that in!

The implementation and existing tests seem perfect to me, only have some minor polish comments :) Making this an approval as at this point I consider any further polishing a bonus, the feature seems ready to benefit xDSL in itself!

Thanks a lot!

File: xdsl/irdl/declarative_assembly_format.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (alexarice) [reply xdsl/irdl/declarative_assembly_format.py:737]: Sorry I wasn't clear, I meant that I think `parse` gets a default implementation from the definition of `parse_optional`. I think you could concoct a weird example where you could still use a region in an optional group even if it's non optional

[the remark this answers is not in the document]
Comment (kimxworrall) [reply xdsl/irdl/declarative_assembly_format.py:737]: You're right, it does get a default implementation

[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py:737]: > Sorry I wasn't clear, I meant that I think `parse` gets a default implementation from the definition of `parse_optional`. I think you could concoct a weird example where you could still use a region in an optional group even if it's non optional

You're right; just for full context: xDSL doesn't implement that case yet (optional group's else-elements) so a lot of other similar declarative variables do not implement it either.

So for the sake of this PR, both removing it or keeping it would be fine :slightly_smiling_face:

If keeping the optional, using the default `parse` implementation if it is working would be nice indeed!

File: tests/irdl/test_declarative_assembly_format.py (hunk 16 above)
Comment (PapyChacal) [reply tests/irdl/test_declarative_assembly_format.py:1285]: My first comment is a bit cryptic;
`OptionalOperandsOp` should contain
`irdl_options = [AttrSizedRegionSegments()]`
to make `irdl_op_definition` happy with the op.

Allowing to test if the declarative syntax is indeed emitting its own error in that case!

**Review by alexarice — APPROVED** [review]
I've had a little test locally and this seems reasonable. Thanks for implementing this. Happy for this to be merged once the remaining comments are resolved

File: xdsl/irdl/declarative_assembly_format.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kimxworrall) [reply xdsl/irdl/declarative_assembly_format.py:737]: Keeping just parse_optional and allowing the default parse implementaion to be used

**Review by superlopuh — APPROVED** [review]

File: tests/irdl/test_declarative_assembly_format.py (hunk 16 above)
Comment (kimxworrall) [reply tests/irdl/test_declarative_assembly_format.py:1317]: Fixed now

File: tests/irdl/test_declarative_assembly_format.py (hunk 17 above)
Comment (kimxworrall) [reply tests/irdl/test_declarative_assembly_format.py:1331]: Have added these two tests
